Fix removing multiple lists

KarolDawidziuk opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
skip
```

## What changes did you make?

After adding a fix for removing multiple lists #4943 there was a problem with removing lists where the first element of the list was not fully selected. This resulted in the entire list item being removed instead of just the selected content. 

The bugged part was in the `isFirstListItem()` function, which checked for a previous item in the list, but did not check if the entire item was selected.

I just added an additional condition in the return statement to check if all the content from the first list item was selected.

## Which issues does your PR resolve?

Closes #5068.

jacekbogdanski reviewed (changes requested):
The issue is a bit more complicated and I can still reproduce an error when deleting the list starting at the first list item and offset 0. It looks like enlarging block elements with visible magic lines incorrectly sets up range offsets, as magicline element is available in the editor content. Once it gets hidden on `saveSnapshot` event we operate on an outdated range when deleting lists content.

I've prepared a simple fix for it by firstly checking for multiple lists case and then preparing the range for deletion (after `saveSnapshot` call). It could be also implemented by extending `magicline` plugin with an additional generic event like `hideUIView` or feature detection.

@KarolDawidziuk please take a look at my proposition (it still may require some refactoring) and fix the remaining issues in this PR.

jacekbogdanski commented on core/editable.js:
```diff
@@ -1681,53 +1682,56 @@
 	}
 
+	function getRangeForSelectedLists( range ) {
+		var isListParent = null,
+			listParent = null,
+			firstRangeEl = range.startContainer,
+			parent = firstRangeEl.getParent();
+
+		// Get the parent of the list.
+			while ( !isListParent ) {
+				var parentName = parent.getName();
+
+				// We need to search for first closest parent list.
+					isListParent = CKEDITOR.tools.array.some( listsTypes, function( listType ) {
+						if ( parentName === listType ) {
+							listParent = parent;
+							return parentName === listType;
+						}
+					} );
+
+				parent = parent.getParent();
+			}
```
This code is very unreadable. I will simplify it in the upcoming commit.

KarolDawidziuk commented on tests/core/editable/manual/removealmostentileliets.md:
```diff
@@ -0,0 +1,14 @@
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, basicstyles, list, resize, undo, sourcearea, elementspath, magicline
+
+1. Select everything contained in `[]` including markers.
+2. Press <kbd>delete</kbd> or <kbd>backspace</kbd>
+
+**Expected**
+
+The content from selected range has been properly removed.
+
+**Unexpected**
+
+All two lists have been removed.
```
Yes, there was a typo in the file name. I'll change that and update the description.

KarolDawidziuk commented on tests/core/editable/manual/removemultiplelistunsupported.html:
```diff
@@ -42,6 +42,20 @@ <h2>#3</span>
+</textarea>
+
 <script>
 	( function() {
```
Yes, but I didn't want to create 4 editors in case the browser wasn't supported. So I'd like to stay with that syntax if you don't mind.

KarolDawidziuk commented on tests/core/editable/manual/removemultiplelistunsupported.html:
```diff
@@ -42,6 +42,20 @@ <h2>#3</span>
+
 <script>
 	( function() {
 		if( !CKEDITOR.env.ie && !CKEDITOR.env.gecko ) {
```
These cases are covered in the unit tests. I didn't want to duplicate the testing process. Still, there is a manual test named `removemultiplelist` that is available in every browser.

KarolDawidziuk commented:
Thank you @jacekbogdanski for your improvements.

It's ready for the next one.

I have added more manual tests for removing multiple lists, including nested lists. Additionally, I've added separate manual tests for removing lists with `magicline` plugin.

jacekbogdanski reviewed (approved):
Let's make sure that everything is alright before merging this PR with @ckeditor/qa-team help. It would be good to verify these tests and play a bit with lists:
```
tests/core/editable/manual/removelistatoffset
tests/core/editable/manual/removelistswithmagicline
tests/core/editable/manual/removemultiplelists
tests/core/editable/manual/removemultiplelistswithmagicline
```

Mgsy commented:
It's still possible to reproduce the issue after applying a link to the text that should be removed:

```
<p>test</p>

<ol>
	<li>tes<a href="http://d">[t1</a></li>
	<li><a href="http://d">test2</a></li>
</ol>

<p><a href="http://d">test3</a></p>

<ol>
	<li><a href="http://d">test4</a></li>
	<li><a href="http://d">test5]</a></li>
</ol>
```

![](https://user-images.githubusercontent.com/20988892/152512195-0a75d565-9ee7-49db-b482-c0b33dac96f3.gif)

dufipl commented:
Not only a link but reproduces also when the background colour is set to deleted text:
![list remove error](https://user-images.githubusercontent.com/32898551/152513326-21051a39-e6fc-4bcd-a831-43d9811b9587.gif)

Mgsy commented:
We have tested it with @dufipl and after the last fix everything works as expected and we didn't notice any regression.

jacekbogdanski commented:
Great to hear that, thank you! We still need to add additional unit test coverage for these changes, but it shouldn't stop us from merging this PR and finishing the testing phase.
